### PR TITLE
Ensure `import urllib3` can succeed even if trio and twisted aren't installed

### PR DESCRIPTION
**Pull request by oliland**
This is a draft at resolving https://github.com/python-trio/urllib3/issues/25

This ensures that we never directly import trio / twisted until we need them. We do this by adding some scary logic to `_backends.__init__`.

Needs tests and feedback on the import mechanism.

### CODE REVIEW

**Comment by oliland** [comment]
To reiterate - this is very much a draft - just looking for feedback on the approach before I go about fixing any broken tests.

File: urllib3/_backends/__init__.py [hunk 1]
```diff
@@ -1,9 +1,56 @@
...
+def available_backends():
+    """
+    Returns a list of backend names that we are able to successfully import.
+    """
+    output = []
+    for backend_name, load_backend_fn in sorted(BACKENDS.items()):
+        try:
+            load_backend_fn()
+            output.append(backend_name)
+        except:  # noqa
```
Comment (pquentin) [inline urllib3/_backends/__init__.py]: You probably want at least `except Exception` in order to avoid catching KeyboardInterrupt exceptions.

Any good reason not to simply use `ImportError`? I guess the import can fail with different errors, but as a user, I would like to know if a dependency is half-installed.

Am I missing something?

File: urllib3/_backends/__init__.py [hunk 2]
```diff
@@ -1,9 +1,56 @@
...
+
+def load_backend(backend_name, backend_args):
+    _backend_args = backend_args or {}
+    if backend_name is None:
+        return load_sync_backend()(**_backend_args)
+    elif backend_name in BACKENDS:
+        return BACKENDS[backend_name]()(**_backend_args)
+    else:
+        valid_backends = ", ".join(sorted(BACKENDS.keys()))
+        raise ValueError("Backend must be one of: {}".format(valid_backends))
```
Comment (pquentin) [inline urllib3/_backends/__init__.py]: Another option would be to use a static list of backends in the `else` clause, and let the loading of a backend fail if the dependency is not installed. Which might be more intuitive?

File: urllib3/_backends/__init__.py (hunk 1 above)
Comment (oliland) [reply urllib3/_backends/__init__.py]: That sounds like a good idea - thanks!

File: urllib3/_backends/__init__.py (hunk 2 above)
Comment (oliland) [reply urllib3/_backends/__init__.py]: I'm not sure I follow - can you rephrase your suggestion?

This else block is raising an error if the user specifies an unknown backend. At this point, I'm assuming that we will have a finite list of backends in urllib3 and will not allow users to write their own drivers without adding them upstream - maybe we want to reconsider this, though.

Comment (pquentin) [reply urllib3/_backends/__init__.py]: Sorry, this was not easy to understand. I'll try to be clearer this time.

So I had two suggestions. The first one, above, was simply about replacing `except` by `except Exception` or `except ImportError`.

This suggestion would have more impact. I was wondering about hard coding the valid backends (currently that would be trio, twisted and sync) and always report them as available in the ValueError raised in the else clause of `load_backend`. I think this would help with the discovery of new backends. Now, if a user tries to use such a backend but the import fails, I don't know if we should catch the ImportError exception or let it propagate.

I'm really thinking aloud here: I don't know if that would be better or not!

Comment (oliland) [reply urllib3/_backends/__init__.py]: @pquentin Apologies - I think I've misled you. The `available_backends` function is not used - the error I am raising contains all "known" backends.

I added `available_backends` to assist in debugging from a repl.

Comment (oliland) [reply urllib3/_backends/__init__.py]: I'm going to remove the `available_backends` function as it serves no purpose beyond debugging, and probably wouldn't be exposed as a user visible API.

Comment (pquentin) [reply urllib3/_backends/__init__.py]: Not sure how I managed to misread this. Sorry!

I think we can leave available_backends as it is.

**Comment by oliland** [comment]
I've added some code that documents the issue in 1a73eff - @pquentin any ideas?

FWIW, I've had a scan of http://python-notes.curiousefficiency.org/en/latest/python_concepts/import_traps.html and immediately spot any bad practices, but could well be missing something.

**Comment by njsmith** [comment]
Regarding the public API, see: https://github.com/python-trio/urllib3/issues/25#issuecomment-385767734

**Comment by oliland** [comment]
TIL something about Python.

This:
```python
from .sync_backend import SyncBackend


def load_sync_backend():
    return SyncBackend
```
Is subtly different to this:
```python
def load_sync_backend():
    from .sync_backend import SyncBackend
    return SyncBackend
```
I believe that in the second case, I believe am triggering a variant of the [Double Import Trap](http://python-notes.curiousefficiency.org/en/latest/python_concepts/import_traps.html#the-double-import-trap), as while the repro is different, the outcome is that I have two copies of the same module.

Two copies of the same module means I have two copies of the `LoopAbort` class, which can never be equivalent as their `type()` is different (and I believe [this code](https://github.com/python/cpython/blob/master/Objects/object.c#L643) would be used to check for Exception equality, as the BaseException class doesn't declare a rich comparison function).

Perhaps reassuringly, this also fails:
```python
from .sync_backend import SyncBackend


def load_sync_backend():
    from .sync_backend import SyncBackend
    return SyncBackend
```

I'll need to think quite deeply about how to solve this, as ideally we would allow ourselves to load backends dynamically, but we will need to acknowledge that they cannot communicate using types (including Exceptions) as their types will be considered different.

**Comment by pquentin** [comment]
Wow, nice detective work here! 💯 

Would adding a rich comparison function to LoopAbort work?

**Comment by njsmith** [comment]
`except` blocks use `isinstance`, not `==`, so I don't think rich comparison will help. I guess some `__subclasscheck__` metaclass hackery would be possible in theory, but... we should figure out the actual problem :-).

This looks *really* weird to me, in particular this bit:
```
<class 'urllib3._backends._common.LoopAbort'> 140625114597352
<class 'urllib3._backends._common.LoopAbort'> 140625077300856
```
I really don't think this is supposed to be possible!

From the double-import trap link: "The reason this is problematic is that every module in that directory is now potentially accessible under two different names", but in the above, the two different `LoopAbort` objects both think that they're in the `urllib3._backends._common` module, and if they have the same name then it's supposed to mean that it has just one entry in `sys.modules` and only gets imported once. And importing inside a function or at the top-level *shouldn't* change this...

@oliland I don't suppose you've cut down a minimal reproducer?

To pick some random folks who I know know a lot about the import system: @ncoghlan @brettcannon Do you have any idea how in Python 3.6 you could end up with two different imports of the same module under the same name?

**Comment by njsmith** [comment]
Here's another oddity: when I add a `print` statement to `urllib3/_backends/_common.py`, it only seems to execute once, suggesting that it actually only gets imported once. Yet I can still reproduce the issue. So where's this other `LoopAbort` even coming from?

I thought this must mean that there's another copy of `_common.py` somehow on the PYTHONPATH, like the one in the `build/` directory or something, but when I ran `strace -e file -f pytest ...`, then grepped the output for `_common`, I get:

```
4070  stat("/home/njs/src/urllib3/urllib3/_backends/_common.py", {st_mode=S_IFREG|0644, st_size=1069, ...}) = 0
4070  stat("/home/njs/src/urllib3/urllib3/_backends/_common.py", {st_mode=S_IFREG|0644, st_size=1069, ...}) = 0
4070  openat(AT_FDCWD, "/home/njs/src/urllib3/urllib3/_backends/__pycache__/_common.cpython-35.pyc", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
4070  openat(AT_FDCWD, "/home/njs/src/urllib3/urllib3/_backends/_common.py", O_RDONLY|O_CLOEXEC) = 8
4070  stat("/home/njs/src/urllib3/urllib3/_backends/_common.py", {st_mode=S_IFREG|0644, st_size=1069, ...}) = 0
4070  stat("/home/njs/src/urllib3/urllib3/_backends/_common.py", {st_mode=S_IFREG|0644, st_size=1069, ...}) = 0
4070  openat(AT_FDCWD, "/home/njs/src/urllib3/urllib3/_backends/__pycache__/_common.cpython-35.pyc", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
4070  openat(AT_FDCWD, "/home/njs/src/urllib3/urllib3/_backends/_common.py", O_RDONLY|O_CLOEXEC) = 17
4070  stat("/home/njs/src/urllib3/urllib3/_backends/_common.py", {st_mode=S_IFREG|0644, st_size=1069, ...}) = 0
4070  lstat("/home/njs/src/urllib3/urllib3/_backends/_common.py", {st_mode=S_IFREG|0644, st_size=1069, ...}) = 0
```

If we ignore the `stat`s and focus on the `openat`s, we see that this python process is in fact opening `/home/njs/src/urllib3/urllib3/_backends/_common.py` twice, and no other files named `_common.py` at all.

So where the heck does this other `LoopAbort` even come from?

**Comment by ncoghlan** [comment]
My initial suspicion would be that you might be running into one of the subtle race conditions [1] that have affected the per-module import locks at various points in time, and are hence getting two concurrent imports of the "urllib3._backends._common" module. However, that would only be a plausible explanation if the error was coming up on 3.6.2, and everything worked fine on 3.6.3 and later.

[1] https://bugs.python.org/issue30891, https://bugs.python.org/issue30876, https://bugs.python.org/issue31070

**Comment by oliland** [comment]
@njsmith good points. As you've pointed out that both imports should resolve to the same key in `sys.modules`, I've unhidden my initial findings as they are not fully resolved.

I've really struggled to repro this outside of the urllib3 repository. The only other thing I've noticed is that only the `id` of the `LoopAbort` class is different, the `id`'s of `sys.modules["urllib3._backends._common"]` appears to be the same.

There's tons of other weird stuff, hard to tell if relevant or not. Adding another line to `consume_bytes` just above `raise LoopAbort` appears to fix the problem:
```
context['h11_response'] = event
from .._backends._common import LoopAbort
raise LoopAbort
```
Which is cool, apart from the fact that this shouldn't be necessary. For bonus points, add these lines:
```
context['h11_response'] = event
print("about to raise loopabort")
print(id(LoopAbort))
from ._backends._common import LoopAbort
print(id(LoopAbort))
raise LoopAbort
```
What happens here? You guessed it - an `UnboundLocalError` is thrown - not on the second `print(id())`, but the **first**, which implies that the inline import declaration somehow destroys the outer declaration even before the inline import is executed?!

I'll continue to try to extract a repo, though at this point I think I will need to delete the rest of urllib3 around this code to get there.

**Comment by oliland** [comment]
Having failed to produce a minimal repro, I started to delete the rest of urllib3 around this bug and believe I have found the code responsible for this confusion:

https://github.com/python-trio/urllib3/blob/bleach-spike/test/test_no_ssl.py#L67

Our failing test class in question uses this `ModuleStash` object which appears to mutate `sys.modules`.

On setup it iterates through sys.modules and pops any modules related to urllib3, then re-adds them on teardown.

Removing all calls to this ModuleStash object solves the problem and all imports resolve to the same `id`.

Interestingly, it's possible to make the tests pass 50% of the time if you only remove the line `module_stash.stash()`. Only by removing both `module_stash.stash()` and `module_stash.pop()` is the issue enturely resolved.

There are plenty of caveats in the docs [1][2] around screwing around with sys.modules, so we definitely shouldn't do this if it's purely in the interest of coverage. Python 3 docs come with a particularly suitable warning:

> ...deleting essential items from the dictionary may cause Python to fail.

so tl;dr: It looks like two ingredients are required to get this crazy behaviour:

0. mutating sys.modules
1. imports at runtime instead of import time

[1] https://docs.python.org/3/library/sys.html#sys.modules
[2] https://docs.python.org/3/library/importlib.html#importlib.reload

**Comment by oliland** [comment]
It looks like this crazy module stashing thing is in upstream urllib3, so we should definitely make a case for removing it there.

**Comment by njsmith** [comment]
Wow, yeah, that could certainly produce weird effects. Excellent detective work.

It makes sense to want to test that urllib3 copes with a missing `ssl` module, because that's a thing that happens. But I think the way I'd test it would be to spawn a child process with a clean state, and try importing there.

I agree that it'd be good to try to fix this upstream. While that's happening, we can mark these tests as skipped, so they don't block progress on the branch.

**Comment by njsmith** [comment]
Btw, I think this is demonstrating a weird quirk of how python works in general:

> the inline import declaration somehow destroys the outer declaration even before the inline import is executed?!

When python compiles a function body, it statically looks for any variables that are involved in assignment operations, and marks those as local, throughout the whole function body. This is why if you want to assign to a local from inside a function, you need to say so explicitly:

```python
X = 1
def foo():
    # creates a local, doesn't affect global
    X = 2

def bar():
    # works fine
    print(X)

def baz():
    # fails! Python thinks you're trying to print the local created below!
    print(X)
    X = 2

def quux():
    # tells python that in this function, X refers to the global
    global X
    # now this works
    print(X)
    # and this mutates the global
    X = 2
```

**Comment by ncoghlan** [comment]
Ah, yeah, doing a forced reload can *definitely* cause weirdness, *especially* if other modules are doing "from reloaded.module import attribute" (which is the case for `LoopAbort`) - even if `reloaded.module.attribute` gets replaced, other modules will still have references to the original object.

Very nice detective work tracking that down :)

File: urllib3/_async/connection.py [hunk 3]
```diff
@@ -311,8 +311,7 @@ def __init__(self, host, port, backend=None,
                  source_address=None, tunnel_host=None, tunnel_port=None,
                  tunnel_headers=None):
         self.is_verified = False
-
-        self._backend = backend or SyncBackend()
+        self._backend = backend or load_backend("sync")
```
Comment (oliland) [inline urllib3/_async/connection.py]: Question: if this function constantly calls `from .sync_backend import SyncBackend` - is it constantly re-running the import machinery whenever we create a connection?

If so, we should special case `SyncBackend` to skip the import machinery in the interests of making the common case more efficient.

Alternatively, we should ensure that users have always specified a backend by this point (which is how I plan for async backends to work), and just leave the `or` case for unit tests.

File: urllib3/_backends/__init__.py [hunk 4]
```diff
@@ -1,9 +0,0 @@
-from urllib3.packages import six
```
Comment (oliland) [inline urllib3/_backends/__init__.py]: I don't really see the point in this file any more, so I truncated it. I guess it would make sense to put `load_backend` in here, and have that be the only interface to this module, but that feels like making a decision. 

Additionally, I'm not sure why a private module (`_backends`) would expose a public interface (via `__all__`)?

File: urllib3/_backends/__init__.py [hunk 5]
```diff
@@ -1,9 +0,0 @@
...
-from .sync_backend import SyncBackend
-
-__all__ = ['SyncBackend']
-
-if six.PY3:
```
Comment (oliland) [inline urllib3/_backends/__init__.py]: FWIW, this doesn't seem correct - we should be targeting 3.5+, not 3+ (although maybe twisted is a special case)

File: urllib3/_backends/_loader.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
+import sys
+
+from ..backends import Backend as UserSpecifiedBackend
```
Comment (oliland) [inline urllib3/_backends/_loader.py]: I'm trying to call out here that this is a user interface and should be handled appropriately. It's also not a "Backend" - it's a recipe for building one, but our users shouldn't need to think about that.

File: urllib3/_backends/_loader.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+
+def check_for_python_3_5():
```
Comment (oliland) [inline urllib3/_backends/_loader.py]: Is this needed?

File: urllib3/_backends/twisted_backend.py [hunk 8]
```diff
@@ -15,8 +15,8 @@
 
 
 class TwistedBackend:
-    def __init__(self, reactor):
-        self._reactor = reactor
+    def __init__(self, reactor=None):
```
Comment (oliland) [inline urllib3/_backends/twisted_backend.py]: Ideally all of our backends can be initialized without any arguments, where possible.

File: urllib3/backends.py [hunk 9]
```diff
@@ -0,0 +1,9 @@
+class Backend:
```
Comment (oliland) [inline urllib3/backends.py]: This is a user (or downstream repository) interface.

**Comment by njsmith** [comment]
I'll try to take a closer look later, but quick note: currently our async mode requires 3.**6**+, because it uses an async generator. See the check in `urllib3/__init__.py`.

**Comment by njsmith** [comment]
Ideally we should consolidate this check somehow. Maybe the tests should check `hasattr(urllib3, "AsyncPoolManager")`?

File: urllib3/_backends/__init__.py (hunk 4 above)
Comment (njsmith) [reply urllib3/_backends/__init__.py]: If I did that then probably my logic was, this is a private API from the user point of view (they shouldn't mention `_backend` at all), and an opaque module from the point of view of the rest of the internal code (they can use the stuff in `_backend.__all__`, but shouldn't go digging deeper).

This is a pretty fine distinction and not particularly important.

File: urllib3/_async/connection.py (hunk 3 above)
Comment (njsmith) [reply urllib3/_async/connection.py]: Re-importing an already-imported module is pretty cheap, because python immediately notices that it already has the loaded module cached and uses that. So it's like a few dict lookups.

File: urllib3/_backends/__init__.py (hunk 2 above)
Comment (oliland) [reply urllib3/_backends/__init__.py]: Sorry for the back and forth on this. But it didn’t make sense for me to implement available_backends just yet without agreeing on some kind of user interface for it. I’d like to do that separately.

We also need to be careful with that functionality - as it will import all of the backends, it’s probably something that we don’t want to encourage users to call.

File: urllib3/_async/connection.py (hunk 3 above)
Comment (njsmith) [reply urllib3/_async/connection.py]: Ok, also: what we actually want the logic to be here (I think) is:

* If this is the sync version of the code, then the backend must be `None` or `"sync"` or `Backend("sync")`. Anything else is an error.

* If this is the async version of the code, then the backend must be specified, and must be *not* the sync backend.

And... we kind of want to do this check inside all of the classes that take `backend` arguments.

So I suggest adding a `normalize_backend` function that takes a user `backend` argument, and returns a valid `Backend` or else an error.

The trickiest bit is detecting whether we are in sync mode or async mode. Eventually we should add a nice way to do this to the code generation script, but for now one can do:

```python
async def f():
    return None

obj = f()
if obj is None:
    SYNC_MODE = True
else:
    # prevent unawaited coroutine warning
    obj.close()
    SYNC_MODE = False
```

File: urllib3/_backends/_loader.py (hunk 7 above)
Comment (njsmith) [reply urllib3/_backends/_loader.py]: I don't think so. This should be taken care of by the higher level backend checking logic, plus the fact that you can't get into the async code at all unless you're on a new enough python.

**Comment by oliland** [comment]
@njsmith comments addressed. This is getting a little complex - I've tried to make the error handling as simple as possible for now, we can decide later how best to help users to recover from these errors.

**Comment by pquentin** [comment]
@njsmith ping

**Comment by pquentin** [comment]
A few things are missing:

 * fix the tests
 * fix the async demo
 * fix the wording in `trio backend requires urllib3 to be built with async suppor'` errors: urllib3 is always built with sync and async support

**Comment by pquentin** [comment]
Hello! I opened https://github.com/python-trio/urllib3/pull/47 which builds on this PR and addresses the comments.
